Create fulcio-spec.md

tracymiranda opened the pull request:
Move Fulcio spec from Google doc to Github

#### Summary
This is the first Sigstore spec doc being moved to github for public collaboration

#### Release Note
NONE

#### Documentation
This migrates the Google doc with the following modifications:

- Add title and summary
- Separate introduction (why) and overview (what) sections
- Add numbering to sections for ease of referencing sections
- Improve consistency and conciseness of some text

pwelch commented on fulcio-spec.md:
```diff
@@ -0,0 +1,307 @@
+
+#### 2.2.1. OpenID Connect
+
+Identity is provided by the OpenID Connect (OIDC) framework, an authentication extension to OAuth 2.0. Fulcio accepts OIDC ID tokens from a configured list of identity providers (IDPs). Before certificate issuance, Fulcio verifies the signed OIDC ID token, and extracts the identity from the ID token. Fulcio is designed to be easily extensible and can be configured with any IDP that conforms to th ...
```
Thoughts on linking out to OpenID overview doc?

```markdown
Identity is provided by the [OpenID Connect (OIDC) framework](https://openid.net/developers/how-connect-works/),
```

pwelch commented on fulcio-spec.md:
```diff
@@ -0,0 +1,307 @@
+
+Users of Sigstore can verify via cryptographic proof that certificates are included in the log
+
+along with monitoring the log for inconsistencies.
```
[nit] looks like an extra line got added

pwelch commented on fulcio-spec.md:
```diff
@@ -0,0 +1,307 @@
+
+the Tink keyset. The Tink keyset MUST be encrypted with a KMS key, and stored in
+
+a JSON format. Tink keysets use strong security defaults and are the most secure way to store an encryption key locally.
```
[nit] looks like some extra newlines got added on migration

pwelch commented:
Thanks for migrating this!

chinenyeokafor commented on fulcio-spec.md:
```diff
@@ -0,0 +1,307 @@
+
+## 3. Issuance - Life of a Request
+
+The client submits a certificate request to Fulcio. The certificate request MUST contain a certificate request described subsequently and OpenID Connect (OIDC) identity token. This is a signed JWT containing information about the principal (identity of the client), the issuer (who issued the identity token - Google, Microsoft, GitHub, etc.) and additional metadata such as expiration. The principa ...
```
Do you mean "identity of the signer" instead of "identity of the client"? 'client' has been consistently used to refer to the signing/verification tool

haydentherapper replied:
Yes, I like `signer` more, that is more precise. I overload `client` to refer to the signer.

chinenyeokafor commented on fulcio-spec.md:
```diff
@@ -0,0 +1,307 @@
+
+Fulcio does not support MAC-based authentication.
+
+Once the client has been authenticated, Fulcio MUST verify the client is in possession of the private key of the public key they’ve submitted. Fulcio MUST verify the signed challenge or CSR. For a signed challenge, this MUST be a signature over the identity claim of the ID token, which SHOULD be the `sub` claim but MAY be a non-standard claim as Fulcio supports configuration of this claim. The ch ...
```
Is 'token' meant instead of 'client' in 'Once the client has been authenticated'?

haydentherapper replied:
Yes, thanks, `token` would be more accurate than `client` - "Once the token has been authenticated".

tracymiranda commented:
Based on initial reviews I've added the OIDC link and fixed the spacing issues. 
For feedback on the terminology I've opened up issues so we can get reviews on it from the original authors.

tracymiranda commented:
Thanks for the review @haydentherapper, added in the fixes from @chinenyeokafor's review.
